File: DS/DSU.py
class DSU:
    def __init__(self, n):
        self.parent = list(range(n))

    def find(self, target):
        if self.parent[target] != target:
            self.parent[target] = self.find(self.parent[target])
        return self.parent[target]

    def union(self, c, p):
        self.parent[self.find(c)] = self.find(p)

File: DS/DSU.cpp
#include <bits/stdc++.h>

using namespace std;

struct DSU {
    int parent[10];
    DSU() {
        for(int i = 0; i < 10; ++i) {
            parent[i] = i;
        }
    }
    int find(int i) {
        if(parent[i] != i) {
            parent[i] = find(parent[i]);
        }
        return parent[i];
    }
    void combine(int i, int j) {
        parent[find(i)] = parent[j];
    }
    void print() {
        for(int i = 0; i < 10; ++i) {
            cout << parent[i] << " ";
        }
        cout << endl;
    }
};

int main() {
    DSU dsu;
    dsu.print();
    dsu.combine(1, 2);
    dsu.combine(1, 5);
    dsu.print();
    return 0;
}
